### ISSUE TITLE
Struggling to get VMware workstation running on SP5 m3 Ubuntu 13.10

### ISSUE THREAD

**Issue by antomicblitz** [linux-surface/linux-surface#1143]
Hi,

I was hoping someone could let me know the best way to get VMware working on my surface after installing the Linux surface kernel so I can run a virtual machine on my surface pro 5. I am able to get everything working following the installation instructions, except when I try to install VMware, it fails and looks for compiled vmmon and vmnet. I tried to compile these modules using gcc but it didn't seem to work. I am struggling a bit on how to get this working. Does anyone have it working already who can give some pointers?

**Comment by antomicblitz** [comment]
Just want to add that I got it working. I think there might have been some issue with secure boot. When I disabled it, I was able to compile the modules.

**Comment by antomicblitz** [comment]
> There are generally two potential issues with external kernel modules (apart from you having to compile stuff yourself e.g. via DKMS):
> - A kernel version mismatch such that the module won't compile. This is mostly because our kernels are quite a bit newer than what Ubuntu has in their repos. So if you try to install a DKMS package targeted at those Ubuntu kernels, it may not work on ours.
> - Kernel lockdown. This prevents the kernel from loading any unsigned (i.e. untrusted) modules. Meaning you will have to set up a signing key to sign the modules and tell the kernel that all modules signed by that key are to be trusted. DKMS should normally do that for you automatically.
> 

Yes I think these were both issues. I ended up choosing the surface kernel which matched the one used by Ubuntu, turned off secure boot, and then it worked. Now the only issue I have is the camera. Once I get that working, it will be a pretty good machine. Unfortunately, I think the newest Ubuntu may have broken the camera. None of the precompiled modules got the camera to be recognized by cheese, so I went and compiled libcamera myself, which also involved digging up a random ipu-c3.bin file or something from the Linux-firmware Ubuntu package, and recompiling. Although I can open the camera in gcam and it shows up, it still is just a black screen in cheese and I still can't get gstreamer recognized when I compile with clang which is extremely frustrating. But hopefully I can figure it out. If I can share more details about the camera issues, do you think you could help me to get it working? It is the last stupid thing to fix and then it will be a pretty great little computer.

**Comment by qzed** [comment]
Unfortunately I really don't know much about the camera stuff beyond the very basics. As in: I know there's a gstreamer plugin but I don't know what could cause that to only show black. I guess an alternative could be the loopback method if it doesn't work in cheese directly? @kbingham and @djrscally will probably know more about that sort of stuff.

**Comment by antomicblitz** [comment]
Thanks! I know even less haha. I appreciate the help though. It's still awesome even if the camera doesn't work, almost everything else is running very well. This kernel is awesome.

**Comment by antomicblitz** [comment]
Here is the guide I am following g to get the camera working btw
https://neilzone.co.uk/2021/08/working-front-and-rear-cameras-on-debian-11-on-a-surface-pro-6

**Comment by kbingham** [comment]
Surface pro 5 looks like the standard camera setup so it should be operable. 

https://github.com/linux-surface/linux-surface/wiki/Camera-Support

Are you saying it always gives a black image in cheese/qcam too?

Which version of libcamera are you running? If your image is black or did be anything from not capturing frames successfully, to capturing but with incorrect AGC or Sensor driver gain model.... So we'll have some digging to do

**Comment by kbingham** [comment]
Oh sorry I missed that in an earlier reply. It does work in our tools, just not cheese? 
I'm really hoping the pipewire integration will land in browsers soon so you'll get camera integration directly in chromium/Firefox without a loopback.

**Comment by qzed** [comment]
We should probably split this into two issues, one for VMWare and one for the cameras.

**Comment by antomicblitz** [comment]
Okay I just want to give one other update. I decided to ditch Ubuntu and go with Fedora 38 Workstation. To my surprise, everything works, including the camera (although it still doesn't work with most major applications like Zoom or chromium and the quality is not that great). But still, it's good enough. I was also able to build the modules in VMware after running 
`dnf install gcc`
Then, I successfully followed all the instructions in the surface-linux guide for Fedora. I created a new MOK key and signed the kernel modules I compiled for VMware using the following commands:
`openssl req -new -x509 -newkey rsa:2048 -keyout MOK.priv -outform DER -out MOK.der -nodes -days 36500 -subj "/CN=VMware/"`
`sudo /usr/src/kernels/$(uname -r)/scripts/sign-file sha256 ./MOK.priv ./MOK.der $(modinfo -n vmmon)`
`sudo /usr/src/kernels/$(uname -r)/scripts/sign-file sha256 ./MOK.priv ./MOK.der $(modinfo -n vmnet)`
`sudo mokutil --import MOK.der`

I can now run VMware perfectly, so I guess this issue is solved.
